File: src/main/java/com/yatspec/e2e/config/RabbitListenerInterceptorConfig.java
package com.yatspec.e2e.config;

import com.yatspec.e2e.captor.rabbit.ConsumeCaptor;
import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import org.springframework.amqp.core.Message;
import org.springframework.amqp.core.MessageBuilder;
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.context.annotation.Configuration;

import javax.annotation.PostConstruct;

/*
    This config adds the interception of messages to RabbitMq listeners
*/
@ConditionalOnProperty(name = {"yatspec.lsd.db.connectionstring"})
@ConditionalOnBean(SimpleRabbitListenerContainerFactory.class)
@Configuration
@RequiredArgsConstructor
public class RabbitListenerInterceptorConfig {

    private final SimpleRabbitListenerContainerFactory simpleRabbitListenerContainerFactory;
    private final ConsumeCaptor consumeCaptor;

    @PostConstruct
    public void postConstruct() {
        simpleRabbitListenerContainerFactory.setAfterReceivePostProcessors(this::postProcessMessage);
    }

    @SneakyThrows
    private Message postProcessMessage(final Message message) {
        consumeCaptor.captureConsumeInteraction(MessageBuilder.fromMessage(message).build());
        return message;
    }
}

File: rabbitmq/src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/config/RabbitListenerInterceptorConfig.kt
package io.lsdconsulting.lsd.distributed.interceptor.config

import io.lsdconsulting.lsd.distributed.connector.model.InteractionType
import io.lsdconsulting.lsd.distributed.interceptor.captor.rabbit.RabbitCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.rabbit.mapper.deriveExchangeName
import jakarta.annotation.PostConstruct
import lsd.logging.log
import org.springframework.amqp.core.Message
import org.springframework.amqp.core.MessageBuilder
import org.springframework.amqp.core.MessagePostProcessor
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Configuration

/*
    This config adds the interception of messages to RabbitMq listeners
*/
@ConditionalOnProperty(name = ["lsd.dist.connectionString"])
@ConditionalOnBean(SimpleRabbitListenerContainerFactory::class) // TODO What if there is no bean of this type?
@Configuration
class RabbitListenerInterceptorConfig(
    private val simpleRabbitListenerContainerFactory: SimpleRabbitListenerContainerFactory,
    private val rabbitCaptor: RabbitCaptor,
) {

    @PostConstruct
    fun postConstruct() {
        simpleRabbitListenerContainerFactory.setAfterReceivePostProcessors(MessagePostProcessor { message: Message ->
            postProcessMessage(
                message
            )
        })
    }

    private fun postProcessMessage(message: Message): Message {
        try {
            val exchangeName =
                deriveExchangeName(message.messageProperties, message.messageProperties.receivedExchange)
            rabbitCaptor.captureInteraction(
                exchangeName,
                MessageBuilder.fromMessage(message).build(),
                InteractionType.CONSUME
            )
        } catch (t: Throwable) {
            log().error(t.message, t)
        }
        return message
    }
}
